### PR TITLE
Teamsintegration

### PR DESCRIPTION
**Pull request by subashc2023**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -31,6 +31,14 @@ We use these to make it compatible with running the same script from Lambda and
 
 ### Cloud Formation
 If you enter and slack URL and/or SNS it will publish in addition to logging.
+If you do this, you must also do the follow
```
Comment (jimzucker) [inline README.md]: @subashc2023  this needs to be automated in the cloud formation

File: get_forecast.py [hunk 2]
```diff
@@ -51,13 +51,11 @@
 from urllib.error import URLError, HTTPError
 from base64 import b64decode
 
-# Initialize you log configuration using the base class
```
Comment (jimzucker) [inline get_forecast.py:54]: we should not remove

File: get_forecast.py [hunk 3]
```diff
@@ -51,13 +51,11 @@
...
 logging.basicConfig(level = logging.INFO)
 logger = logging.getLogger()
-
```
Comment (jimzucker) [inline get_forecast.py:57]: Please dont remove lines uncessariy

File: get_forecast.py [hunk 4]
```diff
@@ -51,13 +51,11 @@
...
-
```
Comment (jimzucker) [inline get_forecast.py:58]: same

File: README.md (hunk 1 above)
Comment (jimzucker) [reply README.md]: we ahve to add this poilicy for reading secrets

File: get_forecast_cf.yaml [hunk 5]
```diff
@@ -15,7 +21,7 @@ Parameters:
   snsArnParameter:
     Type: String
     Default: ""
-    Description: Enter sns ARN if you want slack sent.
+    Description: Enter sns ARN if you want SNS sent.
```
Comment (jimzucker) [inline get_forecast_cf.yaml]: change to "Enter sns ARN,used to send slack or teams"

File: get_forecast.py [hunk 6]
```diff
@@ -357,8 +386,8 @@ def lambda_handler(event, context):
     try:
         publish_forecast(boto3)
     except Exception as e:
-        print(e)
-        raise Exception("Cannot connect to Cost Explorer with boto3")
+        logger.error("Failed: publish_forecast")
```
Comment (jimzucker) [inline get_forecast.py]: lets discuss, we should print details here so we do have to read the code that is why we print e and make a reasonable exception
 
        print(e)
       raise Exception("Cannot connect to Cost Explorer with boto3")

File: get_forecast.py [hunk 7]
```diff
@@ -357,8 +386,8 @@ def lambda_handler(event, context):
...
+        raise e
```
Comment (jimzucker) [inline get_forecast.py]: we want ot raise a readable exception

File: get_forecast.py [hunk 8]
```diff
@@ -131,19 +143,36 @@ def send_sns(boto3_session, sns_arn, message):
...
+    except Exception as e:
+        logger.info("Disabling Teams, URL not found", e)
 
     try:
-        sns_arn=get_secret(secrets_manager_client, SNS_SECRET_KEY_NAME)
+        sns_arn=json.loads(secret)[SNS_SECRET_KEY_NAME]
         send_sns(boto3_session, sns_arn, message)
     except Exception as e:
         logger.info("Disabling SNS, Arn not found")
-
```
Comment (jimzucker) [inline get_forecast.py:146]: dont delete lines adds to the diff

File: get_forecast.py [hunk 9]
```diff
@@ -131,19 +143,36 @@ def send_sns(boto3_session, sns_arn, message):
...
+    except Exception as e:
+        logger.error("get_secret failed")
+
     try:
-        slack_url='https://' + get_secret(secrets_manager_client, SLACK_SECRET_KEY_NAME)
+        slack_url = json.loads(secret)[SLACK_SECRET_KEY_NAME]
         send_slack(slack_url, message)
     except Exception as e:
-        logger.info("Disabling Slack, URL not found")
+        logger.info("Disabling Slack, URL not found", e)
```
Comment (jimzucker) [inline get_forecast.py]: lets discuss why print e here, if you want it for debugging add 
logger.debug(e) 

but it just makes logs messy

File: get_forecast.py [hunk 10]
```diff
@@ -131,19 +143,36 @@ def send_sns(boto3_session, sns_arn, message):
...
+    if 'SECRETS_MANAGER_ENDPOINT' in os.environ:
+        logger.info("Using secrets_manager_endpoint.")
+        secrets_manager_client = boto3_session.client('secretsmanager', os.environ['SECRETS_MANAGER_ENDPOINT'])
+    else:
+        secrets_manager_client = boto3_session.client('secretsmanager')
+
+    try:
+        secret = get_secret(secrets_manager_client)
+    except Exception as e:
+        logger.error("get_secret failed")
```
Comment (jimzucker) [inline get_forecast.py]: isnt this a fatal error
1. logger.error("get_secret failed") should be logger.debug
2. should this be fatal and raise e?

File: get_forecast.py [hunk 11]
```diff
@@ -131,19 +143,36 @@ def send_sns(boto3_session, sns_arn, message):
 
 
 def display_output(boto3_session, message):
-    secrets_manager_client = boto3_session.client('secretsmanager')
+    secrets_manager_client = None
+    secret = ""
+    if 'SECRETS_MANAGER_ENDPOINT' in os.environ:
+        logger.info("Using secrets_manager_endpoint.")
```
Comment (jimzucker) [inline get_forecast.py]: what does this mean?

File: get_forecast.py [hunk 12]
```diff
@@ -67,31 +65,20 @@
...
+    secret = ""
     try:
-        text_secret_data = ""
-        get_secret_value_response = sm_client.get_secret_value( SecretId=AWSGENIE_SECRET_MANAGER )
-    except ClientError as e:
+        secret = sm_client.get_secret_value( SecretId=AWSGENIE_SECRET_MANAGER )["SecretString"]
+    except Exception as e:
         if e.response['Error']['Code'] == 'ResourceNotFoundException':
-            logger.error("The requested secret " + secret_name + " was not found")
+            logger.error("The requested secret " + secret_key_name + " was not found")
```
Comment (jimzucker) [inline get_forecast.py]: where is secret_key_name defined? a re these error handling still relevant

**Review by jimzucker — CHANGES_REQUESTED** [review]

File: get_forecast.py (hunk 11 above)
Comment (subashc2023) [reply get_forecast.py]: it was from the testmerge branch, but it always takes the else path. Now removed.

File: get_forecast.py [hunk 13]
```diff
@@ -67,31 +67,18 @@
     logger.info("Not running as lambda")
 
 
-def get_secret(sm_client,secret_key_name):
-    # if AWS_LAMBDA_FUNCTION_NAME == "":
```
Comment (jimzucker) [inline get_forecast.py:71]: can we still run and test the script outside lambda w/out this?

**Review by jimzucker — CHANGES_REQUESTED** [review]

**Review by jimzucker — APPROVED** [review]
LGTM

**Pull request merged**